### PR TITLE
!!!FEATURE: Offer resolution strategies when conflicts arise during rebase

### LINKED ISSUE #3744: FEATURE: Offer resolution strategies when conflicts arise during rebase
When conflicts arise during publishing, the user needs a way to decide what to do about them. A possible conflict situation would be for instance:

> `Editor A` removes content node `TextElement-1` and publishes their change. Meanwhile, `Editor B` edits the contents of `TextElement-1`. Once `Editor B` publishes their workspace, `TextElement-1` is already gone. The changes to `TextElement-1` cannot be published.

In this case, `Editor B` would currently run into an exception during publish:
Screenshot_2024-03-25_14-16-37 Publish Error

That is, if `Editor A` had published their removal changes during an ongoing editing session of `Editor B`. In this case, `Editor B` would not be informed about their workspace having gone into `OUTDATED_CONFLICT` state.

When `Editor B` reloads the page, their workspace state is properly reflected:
Screenshot_2024-03-25_15-23-43 Sync Button

Upon clicking the `Synchronize personal workspace` button, `Editor B` is currently informed that there's no way out of here:
Screenshot_2024-03-25_12-57-27 Snyhcronize Workspace Error

`Editor B` is now trapped. Reasonably, they should at least be able to discard their workspace, but this currently leads to an error as well:
Screenshot_2024-03-25_12-58-31 Discard Error

A way to escape this trap would be to perform a `force-rebase` on `Editor B`s workspace. This has been implemented here: https://github.com/neos/neos-development-collection/pull/4939

During the Hamburg spring, @ahaeslich, @mficzel and I discussed the following basic strategies that we can offer the user in case of conflicts:

1. "Discard All" - If publishing fails due to conflicts, we allow the user to throw away all changes
  - This additionally requires an "Are you sure?"-confirmation type dialog
2. "Save what you can" - We allow the user to publish everything that can be published, while throwing away all changes that can't (see: https://github.com/neos/neos-development-collection/pull/4939)
  - For this, we would (ideally in beforehand) display information to the user, so they can understand what they're about to lose if they were to choose this strategy

*Additional strategies had also been discussed (with input from @nezaniel and @mhsdesign), with regards on retaining conflicting changes for later use. We concluded that it's best to focus on the above strategies for 9.0 and explore other solutions later.*

Here's a sketch for a possible dialog workflow:
Untitled-2024-03-26-1410 Rebase Conflict Resolution
(https://excalidraw.com/#room=b7f4a755ffda65849f6a,1SW-iMdJIpdvwWJWwb-CdQ)

***(1)** A Select box in anticipation of `n` possible strategies. At the moment, this select box would contain the options "Discard All", "Save what you can" and "Cancel". The latter will just abort the publishing operation.*
***(2)** Conflicts may be complex and numerous. We will need some way to summarize them for editors.*

**Open Questions**

- [x] What info can we reasonably display about publishing conflicts in the UI?
  - The newly introduced `RebaseFailedException` (see: https://github.com/neos/neos-development-collection/pull/4965) gives us a list of failed commands accompanied by the exception that made the respective command fail. We can deduce from the command which node was affected and what kind of modification was made. We can deduce from the exception class what has caused the conflict.
- [ ] How do we explain that "Discard All" during conflict resolution is not the same thing as "Discard All" in the Publish Dropdown?
  - "Discard All" during conflict resolution will discard everything in the workspace
  - "Discard All" in the Publish Dropdown will discard all changes in the current site, but leave other sites untouched

**Acceptance Criteria**

- [x] If a user clicks on "Synchronize" and a conflict arises during the rebase operation, they are presented with a dialog that allows them to choose between several conflict resolution strategies
- [x] If a user chooses "Cancel" from the resolution strategy selection, the rebase operation is aborted without side effects
- [x] If a user chooses "Discard all" from the resolution strategy selection, their changes are discarded globally
- [x] If a user chooses "Save what you can" from the resolution strategy selection, they are presented with a dialog summarizing what changes they are about to lose
- [x] If a user clicks on "Cancel" in the "Save what you can" summary dialog, they are again presented with the resolution strategy selection
- [x] If a user clicks on "Continue" in the "Save what you can" summary dialog, the rebase operation starts again, this time setting the `force` flag (as established in https://github.com/neos/neos-development-collection/pull/4939) to `true`
- [x] There is an E2E test scenario that creates a conflict state between two editors and asserts the results of choosing "Discard all" as a resolution strategy during rebase
- [x] There is an E2E test scenario that creates a conflict state between two editors and asserts the results of choosing "Save what you can" as a resolution strategy during rebase

### PR DESCRIPTION
**Pull request by grebaldi**
<div align="center">

image

</div>

solves: #3744
builds on: https://github.com/neos/neos-ui/pull/3759

## Description

When editors work in parallel, it may happen that one editor publishes changes that create conflicts with the changes of another editor. Since https://github.com/neos/neos-ui/pull/3665 it has been possible for users update their workspace whenever it has fallen out-of-sync with its respective base workspace. Up until now, this mechanism doesn't handle conflicts. When a rebase operation fails due to conflicts, users are trapped in a workspace that cannot be synced, published nor discarded. The only way out is the CLI (see: https://github.com/neos/neos-development-collection/pull/4939).

**TLDR;** This PR introduces a dialog workflow that allows users to select a resolution strategy for dealing with conflicts that arise during the synchronization of their user workspace.

### How to provoke a conflict

Assuming you have two user accounts with `Neos.Neos:Editor` role (let's call them "editor-a" and "editor-b"):

1. Log in as "editor-a"
2. Create a document "Test"
3. Add a text node to "Test"
4. Publish those changes
5. Edit the text node, without publishing that change
6. Log in as "editor-b" in a private window
  - You should see the orange "Sync" button now
7. Sync "editor-b"'s workspace, so you can see document "Test"
  - This should trigger the synchronization workflow without conflicts
8. Remove document "Test"
9. Publish that change
10. Switch to "editor-a"'s window and reload
  - You should see the orange "Sync" button now
11. Edit the text in the text node on document "Test"
12. Try now to sync "editor-a"'s workspace
  - This should trigger the synchronization workflow with conflicts

### Conflict-Free Synchronization Workflow

<details>
<summary>Show video</summary>

https://github.com/neos/neos-ui/assets/2522299/6f3279a7-5ef6-49d5-af65-98d78bf74f2a

</details>

When the user clicks on the orange "Sync" button, a dialog opens that asks them to confirm their intended action. Once they click "Yes, synchronize now", a process indicator opens that cannot be closed. During the operation a `beforeunload` handler prevents the user from leaving the page. After the operation has finished, the user is presented with a dialog stating the success of the synchronization. Once they click "OK", they see the updated UI state.

If synchronization has removed the document that the user was currently looking at, they have been silently redirected to the closest ancestor of that document.

### Synchronization with Conflicts (Strategy `FORCE`)

<details>
<summary>Show video</summary>

https://github.com/neos/neos-ui/assets/2522299/4f601095-cb48-4643-a7de-1ff3b517dc55

</details>

Again, when the user clicks on the orange "Sync" button, a dialog opens that asks them to confirm their intended action. Once they click "Yes, synchronize now", a process indicator opens that cannot be closed. If the server-side rebase operation fails, the process indicator is interrupted by a dialog that states the presence of conflicts and offers a choice between two strategies to resolve those conflicts:

1. Drop conflicting changes
2. Discard workspace "workspace-name"

If they choose "Drop conflicting changes" as a resolution strategy, the user is again presented with a dialog that asks them to confirm their intended action. Once they click "Yes, drop those changes", the synchronization starts once again, this time transmitting `force: true` to the server-side command handler.

The rest of the workflow is analogous to the Conflict-Free Synchronization Workflow.

#### Conflict List

<details>
<summary>Show screenshot</summary>

image

</details>

The conflict list is shown both in the "Resolution Strategy Selection"-dialog and the "Confirm: Drop conflicting changes"-Dialog. Conflicts are calculated from a list of `CommandThatFailedDuringRebase`-model objects, that carry information about the attempted CR-level command and the exception that prevented this command from being applied.

The conflict list is a vertical list of `<detail>`-elements with a `<summary>` consisting of an annotated node type icon and the label of the affected node. The little orange annotation to the node type icon represents the type of change, which can be one of 4 things:

1. The node has been created
2. The node has been changed
3. The node has been moved
4. The node has been deleted

The list of conflicts has been squashed on the server-side, so only one (the very first) change per node aggregate will be displayed.

#### Conflict Details

<details>
<summary>Show screenshot</summary>

image

</details>

Opening any item in the conflict list will show the user details about that particular conflict. They will once again see what happened to the affected node in their workspace, and then what is the reason that prevents that change from being applied. Additionally, they'll see what document and what site the affected node belongs to (because conflicts don't necessarily pertain to the site the user is currently looking at).

The "reason for conflict" info is a best-effort guess from the exception that had prevented the respective command from being applied. At the moment, the only reason that will be displayed is "[Node] or one of its ancestor nodes has been deleted". It is possible however for this field to remain empty. Over time we can add more scenarios to the server-side handling of conflicts to better describe each situation.

### Synchronization with Conflicts (Strategy `DISCARD_ALL`)

<details>
<summary>Show video</summary>

https://github.com/neos/neos-ui/assets/2522299/d55c9730-fa63-4a7f-aa38-2036629abde4

</details>

Once again, when the user clicks on the orange "Sync" button, a dialog opens that asks them to confirm their intended action. After they click "Yes, synchronize now", a process indicator opens that cannot be closed. Upon failure of the server-side rebase operation, they are presented with a "Resolution Strategy Selection" dialog.

If they choose "Discard workspace ..." as a resolution strategy, the user is again presented with a dialog that asks them to confirm their intended action. When they click "Yes, discard", their entire workspace will be discarded.

The synchronization process is resumed afterwards.

### Error Handling

Errors are handled just like in https://github.com/neos/neos-ui/pull/3759.

## Remaining TODOs

- [x] Convert WorkspaceSync container to typescript
- [x] Convert SyncWorkspaceDialog container to typescript
- [x] Create sub-dialogs for conflict resolution workflow
  - [x]  Split out ConfirmationDialog
  - [x] Dialog for resolution strategy selection
  - [x] Confirmation dialog for "Discard All"
  - [x] Confirmation dialog for "Save what you can" a.k.a. "Force Rebase"
  - [x] Result Dialog
  - [x] Process Indicator
- [x] Move `UI.SyncWorkspaceModal` redux state to `CR.Syncing`
- [x] Wire up new dialogs with `CR.Syncing` state
- [x] Implement DISCARD_ALL strategy through integration with existing Publish/Discard workflow
- [x] Reload Document Tree and Content Canvas after successful rebase
- [x] Add `beforeunload` handler during rebase
- [x] Fix: Warning: Failed prop type: Invalid prop `focusedValue` of type `number` supplied to `SelectBox_ListPreviewFlat`
- [x] Fix: Warning: Each child in a list should have a unique "key" prop. Check the render method of `ConflictList`.
- [x] Fix: Could not find icon `{ prefix: "fas", iconName: "questionmark" }`
- [x] Squash similar conflicts into one to reduce visual redundancy
- [x] Add E2E test scenario that creates a conflict state between two editors and asserts the results of choosing "Discard all" as a resolution strategy during rebase
- [x] Add E2E test scenario that creates a conflict state between two editors and asserts the results of choosing "Save what you can" as a resolution strategy during rebase
- [x] Remove legacy `WorkspaceStatus.OUTDATED_CONFLICT` from ts-interfaces
- [x] Remove legacy `state.ui.remote.isSyncing`
- [x] Remove legacy i18n labels
  - [x] `Neos.Neos.Ui:Main:workspaceSynchronizationApplied`
  - [x] `Neos.Neos.Ui:Main:syncPersonalWorkSpaceConfirm`
  - [x] `Neos.Neos.Ui:Main:syncPersonalWorkSpaceMessage`
  - [x] `Neos.Neos.Ui:Main:syncPersonalWorkSpaceMessageOutdated`
  - [x] `Neos.Neos.Ui:Main:syncPersonalWorkSpaceMessageOutdatedConflict`
- [x] Fix: All 0 change(s) in workspace "user-admin" were sucessfully discarded.

### CODE REVIEW

**Comment by mhsdesign** [comment]
Todos

- [x] fix https://github.com/neos/neos-ui/pull/3759/files#r1575121070

**Comment by bwaidelich** [comment]
I lack the energy and brain to fully review this tonight. But I want to say that this is the best PR that I have ever seen.. Feature wise, but especially its description with screenshots and collapsed videos and all that swag <3

**Comment by dlubitz** [comment]
Ohh wow, this looks really great! Thank you @grebaldi!

File: Classes/Application/SyncWorkspace/ConflictsBuilder.php [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+            $result ??= $coveredDimensionSpacePoint;
+        }
+
+        return $result;
+    }
+
+    private function createIconLabelForNode(Node $node): IconLabel
+    {
+        return new IconLabel(
+            icon: $node->nodeType?->getConfiguration('ui.icon') ?? 'questionmark',
```
Comment (nezaniel) [inline Classes/Application/SyncWorkspace/ConflictsBuilder.php]: As we want to get rid of Node::nodeType and we have the ContentRepository available, we can already fetch the nodeType from there

**Review by nezaniel — COMMENTED** [review]
Just one little thing, the rest looks nice 👍

File: Classes/Application/SyncWorkspace/ConflictsBuilder.php (hunk 1 above)
Comment (bwaidelich) [reply Classes/Application/SyncWorkspace/ConflictsBuilder.php]: Very good point. Even if we skip NodeTypeManager overhaul for 9.0 I would really like to get rid of Node.nodeType because that will allow us to serialize the nodes (e.g. allowing us to solve the "higher level hook" to use a job queue)

Comment (grebaldi) [reply Classes/Application/SyncWorkspace/ConflictsBuilder.php]: Adjusted :)

Comment (nezaniel) [reply Classes/Application/SyncWorkspace/ConflictsBuilder.php]: Nice 👍

**Review by mhsdesign — APPROVED** [review]
Thank you so much ❤️

**Review by kitsunet — APPROVED** [review]
Looks splendid <3 Thank you so much!

**Comment by mhsdesign** [comment]
The new e2e tests are sadly a bit flaky in ci :O
